source/elements/oneMKL/source/domains/spblas/data_types/create_coo_matrix.rst

```diff
@@ -0,0 +1,241 @@
+.. SPDX-FileCopyrightText: 2024 Intel Corporation
+..
+.. SPDX-License-Identifier: CC-BY-4.0
+
+.. _onemkl_sparse_create_coo_matrix:
+
+create_coo_matrix
+=================
+
+Initializes a ``matrix_handle_t`` object with the provided COO data.
+
+.. rubric:: Description and Assumptions
+
+The ``oneapi::mkl::sparse::create_coo_matrix`` function initializes the
+``matrix_handle_t`` object with the provided data.
+
+In the case of buffers, the reference count of the provided buffer is
+incremented which extends the lifetime of the underlying buffer until the
+``smhandle`` is destroyed with ``destroy_sparse_matrix`` or the data is reset with
+``set_coo_matrix_data``.
+
+In the case of USM, the object does not take ownership of the data.
+
```

Comment:
Why would the reference count of the `sycl::buffer` increase given that the API takes in a `buffer` by reference and does not make a copy, see [this link](https://stackoverflow.com/questions/36336899/should-passing-an-object-using-its-reference-increase-the-reference-count-in-a-r#:~:text=Typically%20when%20a%20class%20has,reference%20counter%20does%20not%20increase.)?

In both `USM` and `sycl::buffer` cases, the user and not the library owns the data. Therefore, calling `destroy_sparse_matrix` does nothing directly to `sycl::buffer`s, just as it does nothing to `USM` pointers, except that it may wait for events to complete.

In `buffer` case, users may destroy the buffers before calling `destroy_sparse_matrix` too (although we'd not recommend it), so long as they don't submit any events on the handle+queue after the buffer has been destroyed; this works because buffer destruction is a blocking call.

Reply:
The buffers are given as reference but the implementation for the cuSPARSE and rocSPARSE backends will need to make a copy of the buffer in each handle to be able to get accessors from them and let the SYCL runtime know which buffers is used when submitting an operation. I assume the MKL backend needs to something similar internally?
That means the user is free to destroy the buffer object after it was used to create a handle object but the underlying buffer won't be destroyed until the handle is destroyed.

I will avoid the use of `smhandle` here as you suggested.

Reply:
I've avoided the use of variable names in high level descriptions in https://github.com/oneapi-src/oneAPI-spec/pull/522/commits/c61ff7e65654ee7e0c9a59011b84b8f1f6bc7f26

Reply:
> ... the cuSPARSE and rocSPARSE backends will need to make a copy of the buffer in each handle to be able to get accessors from them and let the SYCL runtime know which buffers is used when submitting an operation. I assume the MKL backend needs to something similar internally? ...

Intel oneMKL stores the USM pointer passed to the `sparse::set_csr_data()` API in its `matrix_handle`. In case of `sycl::buffer`, Intel oneMKL stores the address of the `sycl::buffer` and does not make a copy of that entire buffer. Users should not destroy the `sycl::buffer`s they have passed to oneMKL until they are done setting up all the API calls with the matrix handle. `sycl::buffer` destruction is synchronous so it automatically forces all events using accessors on the buffer to wait, so if users destroy the buffer, then all oneMKL calls before that point will complete. After that destruction, users should not use that `matrix_handle` because the buffer address it still has stored is no longer valid.

Can you post a snippet or a link (maybe from some other domain) to what dispatching to CUDA backend with `sycl::buffer`s looks like? Now that I think of it, I don't know how it would work because (a) cuSPARSE accepts only pointers, (b) creating a host_accessor and passing that pointer to cuSPARSE should/would not work because that would be host data, and (c) (not sure about this, but) you can't create a device accessor on the host and pass the underlying pointer to that to the cuSPARSE backend. Do you know if (c) is possible?

Reply:
We basically use option (c) in BLAS, see [here](https://github.com/oneapi-src/oneMKL/blob/develop/src/blas/backends/cublas/cublas_level3.cpp#L33-L55).

Reply:
Looks like it's somewhere between (b) and (c). It appears there are accessors created for [`onemkl_cublas_host_task` which is just a `host_task`](https://github.com/oneapi-src/oneMKL/blob/3339418fa1318377c65815e6f85aced9efa46e9e/src/blas/backends/cublas/cublas_task.hpp#L60-L71), followed by some [interoperability magic](https://github.com/oneapi-src/oneMKL/blob/3339418fa1318377c65815e6f85aced9efa46e9e/src/blas/backends/cublas/cublas_scope_handle.hpp#L99-L102) to get device pointers which are passed to cuBLAS APIs.

I always assumed that creating accessors for a `host_task` submission would trigger a device-to-host copy of data held by `sycl::buffer`s. Does that not happen for the cuBLAS dispatch, even if you are eventually using just the data pointed to by the device pointers alongside in the cuBLAS API being dispatched?

Reply:
> `sycl::buffer` destruction is synchronous so it automatically forces all events using accessors on the buffer to wait, so if users destroy the buffer, then all oneMKL calls before that point will complete.

That is not quite accurate, buffers are ref-counted and only when the last buffer is destroyed the synchronization and copy occurs.
 From the spec:
> Buffers are reference-counted. When a buffer value is constructed from another buffer, the two values reference the same buffer and a reference count is incremented. When a buffer value is destroyed, the reference count is decremented. Only when there are no more buffer values that reference a specific buffer is the actual buffer destroyed and the buffer destruction behavior defined below is followed.

Storing the address of a buffer goes against this idea. I think the close-source libraries should store copies of the buffers.

You can find my prototype for using cuSPARSE [here](https://github.com/Rbiessy/oneMKL/pull/1) (focusing on trsv only). All of the logic does not happen in a single place but it also works like the BLAS domain. One important difference is that the buffer is provided when [creating a handle](https://github.com/Rbiessy/oneMKL/pull/1/files#diff-7422afbb6342ddcb2974b46ee101515f9e6c737d5601ac32e1b8f2cdfeb622d5R187), at this point the pointer is retrieved from the buffer via the host_task and used with the backend. Later when the handle is used an [accessor is created again](https://github.com/Rbiessy/oneMKL/pull/1/files#diff-4520860f5bfcef6df30df555408192906db7a83056c4fd75018d5f99b7de4e6dR81) from the buffer to ensure that the dependencies are satisfied. We have discussed internally that this is not 100% conformant yet but this works because we will always get the same pointer from a buffer. I can't think of another way to support buffers currently.

Reply:
I've created https://github.com/oneapi-src/oneAPI-spec/issues/524 to discuss this.
I think we can resolve this comment as we've clarified the reasoning. If we decide to use buffer by reference we will also update the documentation so that it does not mention increasing a reference counter.

Reply:
We are discussing this within oneMKL and someone voiced a concern about reference count being host side atomics which could affect performance.  (They had noticed it with regards to sycl::queue being passed by reference vs by value in a for loop of submissions -- causing bubbles and gaps in the kernel execution that were noticeable) If we go ahead and use pass-by-value and discover that the atomic increment of reference count does affects performance, we can capture by value once at top level and then internally pass by reference...

Reply:
I see, I didn't know that. We know there are other performance issues with using buffers so we might as well prioritize making them easier to use. I'd rather not make the change for the sparse domain but I don't think it's that important either way.

Reply:
of course it feels like it should likely be a bug in some level of compiler or driver ... that could be resolved if performance issues arise, but there are also other techniques that could be employed internally (for instance Peter suggested that we could actually storing a copy of the passed-by-reference sycl::buffer internally created with new (which would increment the reference count) and then use that copy everywhere internally ...

Reply:
I think this discussion can be resolved?

Reply:
I had only removed the reference from the workspace buffers. I have removed the references for the other buffers in https://github.com/oneapi-src/oneAPI-spec/pull/522/commits/16ed109830536667b3939cbf6991f6cd726ed322 for the same reasons.

Reply:
OK, I think we all agree internal to oneMKL that it is fine to remove the references for buffers.

Reply:
if we see any performance issue later on due to reference count, we could always use references to buffers inside library internal calls and have reference counting just at top level iface.